boot/initramfs.go

```diff
@@ -83,3 +86,26 @@ func EnsureNextBootToRunMode(systemLabel string) error {
 	}
 	return bl.SetBootVars(m)
 }
+
+// initramfsReboot triggers a reboot from the initramfs immediately
+var initramfsReboot = func() error {
+	if osutil.IsTestBinary() {
+		panic("initramfsReboot must be mocked in tests")
+	}
+
+	out, err := exec.Command("/bin/reboot").CombinedOutput()
+	if err != nil {
+		return osutil.OutputErr(out, err)
+	}
+	// should be unreachable
```

Comment:
Ah yes good point we should use the -f

Reply:
well actually in my testing with an initramfs shell, `-f` isn't necessary in order for the machine to be rebooted immediately without returning from this process, and the main difference when we run with `-f` is that systemd will _not_ unmount things and other more "safe" shut down things when we don't provide `-f`. See:

```
You are in emergency mode. After logging in, type "journalctl -xb" to view
system logs, "systemctl reboot" to reboot, "systemctl default" or "exit"
to boot into default mode.


BusyBox v1.30.1 (Ubuntu 1:1.30.1-4ubuntu6) built-in shell (ash)
Enter 'help' for a list of built-in commands.

# /sbin/reboot; echo after we return
[  OK  ] Stopped target Local Encrypted Volumes.
[  OK  ] Stopped target Emergency Mode.
[   18.061118] [systemd  OK  [1]: ] Stopped target Local Encrypted Volumes.Stopped target Paths.

[   18.062027] [  OK  ] Stopped target Slices.systemd
[1]: Stopp[  OK  ] Stopped Apply Kernel Variables.
[  OK  ] Stopped Load Kernel Modules.[1]: 
Stopped target Paths.
[  OK  ] Stopped Create Static Device Nodes in /dev.
[  OK  ] Stopped Update is Completed.
[   18.067540] systemd[1]: Stopped target Slices.[
  OK  ] Stopped Rebuild Journal Catalog.
[   18.068344] systemd[1]: Stopped target Sockets.[
  OK  ] Stopped Create Volatile Files and Directories.
[  OK  ] Stopped target Local File Systems.
[   18.069730] systemd[1]: Stopped target Timers.
         Unmounting /run/mnt/data...
[   18.071846] systemd[1]: Stopping Emergency Shell...
         Unmounting /run/mnt/kernel...
[   18.072704] systemd[1]: systemd-sysctl.service: Succeeded.
         Unmounting /run/mnt/snapd...
[   18.074080] systemd[1]: Stopped Apply Kernel Variables.
         Unmounting /sysroot...
[   18.075436] systemd[1]: systemd-modules-load.service: Succeeded.
[   18.076428] systemd[1]: Stopped Load Kernel Modules.
         Unmounting Temporary Directory (/tmp)...
[   18.077197] systemd[1]: systemd-tmpfiles-setup-dev.service: Succeeded.
[  OK  ] Stopped Create System Users.
[   18.077785] systemd[1]: Stopped Create Static Device Nodes in /dev.
[   18.079046] systemd[1]: systemd-update-done.service: Succeeded.
[  OK  ] Stopped Emergency Shell.
[   18.079647] systemd[1]: Stopped Update is Completed.
[  OK  ] Unmounted /run/mnt/data.
[   18.080769] systemd[1]: systemd-journal-catalog-update.service: Succeeded.
[   18.081882] systemd[1]: Stopped Rebuild Journal Catalog.
[  OK  ] Unmounted /sysroot.
[   18.082438] systemd[1]: systemd-tmpfiles-setup.service: Succeeded.
[  OK  ] Unmounted Temporary Directory (/tmp).
[   18.083431] systemd[1]: Stopped Create Volatile Files and Directories.
[   18.084482] systemd[1]: Stopped target Local File Systems.
[  OK  ] Stopped target Swap.
[   18.084961] systemd[1]: Unmounting /run/mnt/data...
         Unmounting /run/mnt/base...
[   18.085831] systemd[1]: Unmounting /run/mnt/kernel...
[   18.086652] systemd[1]: Unmounting /run/mnt/snapd...
[   18.087127] systemd[1]: Unmounting /sysroot...
[   18.087729] systemd[1]: Unmounting Temporary Directory (/tmp)...
[   18.088264] systemd[1]: systemd-sysusers.service: Succeeded.
[   18.088766] systemd[1]: Stopped Create System Users.
[   18.089216] systemd[1]: emergency.service: Succeeded.
[   18.089688] systemd[1]: Stopped Emergency Shell.
[   18.090108] systemd[1]: run-mnt-data.mount: Succeeded.
[   18.090558] systemd[1]: Unmounted /run/mnt/data.
[   18.090990] systemd[1]: sysroot.mount: Succeeded.
[   18.091420] systemd[1]: Unmounted /sysroot.
[   18.091806] systemd[1]: tmp.mount: Succeeded.
[   18.092214] systemd[1]: Unmounted Temporary Directory (/tmp).
[   18.092731] systemd[1]: Stopped target Swap.
[   18.093166] systemd[1]: Unmounting /run/mnt/base...
[   18.103914] systemd[1]: run-mnt-snapd.mount: Succeeded.
[  OK  ] Unmounted /run/mnt/snapd.[   18.104600] 
systemd[1]: Unmounted /run/mnt/snapd.
[   18.148163] systemd[1]: run-mnt-kernel.mount: Succeeded.
[  OK  ] Unmounted /run/mnt/kernel.[   18.149070] 
systemd[1]: Unmounted /run/mnt/kernel.
[  OK  ] [   18.150437] Unmounted /run/mnt/base.systemd
[1]: run-mnt-base.mount: Succeeded.
         Unmounting /run/mnt/ubuntu-seed...
[   18.151803] systemd[1]: Unmounted /run/mnt/base.
[   18.152989] systemd[1]: Unmounting /run/mnt/ubuntu-seed...
[   18.161723] systemd[1]: run-mnt-ubuntu\x2dseed.mount: Succeeded.
[  OK  ] Unmounted /run/mnt/ubuntu-seed.
[   18.162547] systemd[1]: Unmounted /run/mnt/ubuntu-seed.
[  OK  ] Reached target Unmount All Filesystems.[   18.163799] 
systemd[1]: Reached target Unmount All Filesystems.
[   18.165105] systemd[1]: systemd-fsck@dev-disk-by\x2dlabel-ubuntu\x2dseed.service: Succeeded.
[  OK  [   18.165202] ] systemdStopped File System Check …dev/disk/by-label/ubuntu-seed.[1]: 
Stopped File System Check on /dev/disk/by-label/ubuntu-seed.
[  OK  [   18.165874] ] systemdStopped File System Check on /dev/vda2.[1]: 
systemd-fsck@dev-vda2.service: Succeeded.
[   18.168816] systemd[1]: Stopped File System Check on /dev/vda2.
[[   18.168899]   OK  systemd] [1]: Removed slice system-systemd\x2dfsck.slice.Removed slice system-systemd\x2dfsck.slice.

[  OK  [   18.169708] ] systemdReached target Shutdown.[1]: 
Reached target Shutdown.
[   18.171765] systemd[1]: Reached target Final Step.
[  OK  ] Reached target Final Step.
[   18.173061] systemd[1]: systemd-reboot.service: Succeeded.
[  OK  ] Finished Reboot.[   18.173176] 
systemd[1]: Finished Reboot.
[  OK  [   18.173719] ] systemdReached target Reboot.[1]: 
Reached target Reboot.
[   18.175277] systemd[1]: Shutting down.
[   18.361907] systemd-shutdown[1]: Syncing filesystems and block devices.
[   18.363997] systemd-shutdown[1]: Sending SIGTERM to remaining processes...
[   18.367810] systemd-journald[180]: Received SIGTERM from PID 1 (systemd-shutdow).
[   18.388079] systemd-shutdown[1]: Sending SIGKILL to remaining processes...
[   18.393952] systemd-shutdown[1]: Unmounting file systems.
[   18.396111] [397]: Remounting '/' read-only in with options 'size=4053428k,nr_inodes=1013357'.
[   18.398032] systemd-shutdown[1]: All filesystems unmounted.
[   18.399293] systemd-shutdown[1]: Deactivating swaps.
[   18.400469] systemd-shutdown[1]: All swaps deactivated.
[   18.401646] systemd-shutdown[1]: Detaching loop devices.
[   18.403438] systemd-shutdown[1]: All loop devices detached.
[   18.404652] systemd-shutdown[1]: Detaching DM devices.
[   18.405888] systemd-shutdown[1]: All DM devices detached.
[   18.407022] systemd-shutdown[1]: All filesystems, swaps, loop devices and DM devices detached.
[   18.409141] systemd-shutdown[1]: Syncing filesystems and block devices.
[   18.410209] systemd-shutdown[1]: Rebooting.
[   18.411789] Unregister pv shared memory for cpu 0
[   18.411800] Unregister pv shared memory for cpu 3
[   18.411802] Unregister pv shared memory for cpu 1
[   18.411806] Unregister pv shared memory for cpu 2
[   18.411808] Unregister pv shared memory for cpu 7
[   18.411810] Unregister pv shared memory for cpu 4
[   18.411812] Unregister pv shared memory for cpu 6
[   18.422375] Unregister pv shared memory for cpu 5
[   18.425477] reboot: Restarting system
[   18.426789] reboot: machine restart
```

or with `-f`:
```
You are in emergency mode. After logging in, type "journalctl -xb" to view
system logs, "systemctl reboot" to reboot, "systemctl default" or "exit"
to boot into default mode.


BusyBox v1.30.1 (Ubuntu 1:1.30.1-4ubuntu6) built-in shell (ash)
Enter 'help' for a list of built-in commands.

# /sbin/reboot -f; echo after
[   10.568900] Unregister pv shared memory for cpu 3
Rebooting.
[   10.568905] Unregister pv shared memory for cpu 4
[   10.568907] Unregister pv shared memory for cpu 5
[   10.568921] Unregister pv shared memory for cpu 7
[   10.568924] Unregister pv shared memory for cpu 6
[   10.568927] Unregister pv shared memory for cpu 1
[   10.568929] Unregister pv shared memory for cpu 0
[   10.578478] Unregister pv shared memory for cpu 2
[   10.581923] reboot: Restarting system
[   10.583708] reboot: machine restart
```

and actually since the emergency shell recommends it, I also tried with `systemctl reboot`:

```
You are in emergency mode. After logging in, type "journalctl -xb" to view
system logs, "systemctl reboot" to reboot, "systemctl default" or "exit"
to boot into default mode.


BusyBox v1.30.1 (Ubuntu 1:1.30.1-4ubuntu6) built-in shell (ash)
Enter 'help' for a list of built-in commands.

# systemctl reboot; echo after
[  OK  ] Stopped target Local Encrypted Volumes.
[  OK  ] Stopped target Emergency Mode.
[   14.959360] systemd[1]: [Stopped target Local Encrypted Volumes.  OK  
] Stopped target Paths.
[   14.960166] [systemdystemd] Stopped Apply Kernel Variables.
[1]: Stopped target Sockets.
[  OK  ] Stopped Load Kernel Modules.
[   14.963276] systemd[1]: [Stopped target Timers.  OK  
] Stopped Create Static Device Nodes in /dev.
[   14.963869] systemd[1]: Stopping Emergency Shell...
[  OK  ] Stopped Update is Completed.
[   14.966550] systemd[1]: systemd-sysctl.service: Succeeded.
[  OK  ] Stopped Rebuild Journal Catalog.
[   14.967910] systemd[1]: Stopped Apply Kernel Variables.
[  OK  ] Stopped Create Volatile Files and Directories.
[  OK  ] Stopped target Local File Systems.
[   14.969229] systemd[1]: systemd-modules-load.service: Succeeded.
         Unmounting /run/mnt/data...
[   14.971959] systemd[1]: Stopped Load Kernel Modules.
         Unmounting /run/mnt/kernel...
[   14.973072] systemd[1]: systemd-tmpfiles-setup-dev.service: Succeeded.
         Unmounting /run/mnt/snapd...
[   14.974434] systemd[1]: Stopped Create Static Device Nodes in /dev.
         Unmounting /sysroot...
[   14.975657] systemd[1]: systemd-update-done.service: Succeeded.
         Unmounting Temporary Directory (/tmp)...
[   14.976762] systemd[1]: Stopped Update is Completed.[
  OK  ] Stopped Create System Users.
[   14.977302] systemd[1]: systemd-journal-catalog-update.service: Succeeded.
[   14.978785] systemd[1]: Stopped Rebuild Journal Catalog.
[   14.979430] systemd[1]: systemd-tmpfiles-setup.service: Succeeded.
[  OK  ] Stopped Emergency Shell.
[   14.980119] systemd[1]: Stopped Create Volatile Files and Directories.
[   14.981110] systemd[1]: Stopped target Local File Systems.
[  OK  ] Unmounted /run/mnt/data.
[   14.981698] systemd[1]: Unmounting /run/mnt/data...
[   14.982674] systemd[1]: Unmounting /run/mnt/kernel...
[  OK  ] Unmounted /sysroot.
[   14.983177] systemd[1]: Unmounting /run/mnt/snapd...
[  OK  ] Unmounted Temporary Directory (/tmp).
[   14.984066] systemd[1]: Unmounting /sysroot...
[   14.985068] systemd[1]: Unmounting Temporary Directory (/tmp)...
[  OK  ] Stopped target Swap.
[   14.985628] systemd[1]: systemd-sysusers.service: Succeeded.
         Unmounting /run/mnt/base...
[   14.986468] systemd[1]: Stopped Create System Users.
[   14.987247] systemd[1]: emergency.service: Succeeded.
[   14.987764] systemd[1]: Stopped Emergency Shell.
[   14.988300] systemd[1]: run-mnt-data.mount: Succeeded.
[   14.989182] systemd[1]: Unmounted /run/mnt/data.
[   14.989641] systemd[1]: sysroot.mount: Succeeded.
[   14.990099] systemd[1]: Unmounted /sysroot.
[   14.990507] systemd[1]: tmp.mount: Succeeded.
[   14.990933] systemd[1]: Unmounted Temporary Directory (/tmp).
[   14.991477] systemd[1]: Stopped target Swap.
[   14.991904] systemd[1]: Unmounting /run/mnt/base...
[   15.001479] systemd[1]: run-mnt-snapd.mount: Succeeded.
[  OK  ] [   15.002278] Unmounted /run/mnt/snapd.systemd
[1]: Unmounted /run/mnt/snapd.
[   15.039054] systemd[1]: run-mnt-kernel.mount: Succeeded.
[  OK  ] Unmounted /run/mnt/kernel.
[   15.040073] systemd[1]: Unmounted /run/mnt/kernel.
[  OK  ] Unmounted /run/mnt/base.
[   15.041564] systemd[1]: run-mnt-base.mount: Succeeded.
[   15.043031] systemd[1]: Unmounted /run/mnt/base.
         Unmounting /run/mnt/ubuntu-seed...
[   15.043128] systemd[1]: Unmounting /run/mnt/ubuntu-seed...
[   15.053502] systemd[1]: run-mnt-ubuntu\x2dseed.mount: Succeeded.
[  OK  ] Unmounted /run/mnt/ubuntu-seed.
[   15.054141] systemd[1]: Unmounted /run/mnt/ubuntu-seed.
[  OK  ] [   15.055664] Reached target Unmount All Filesystems.systemd
[1]: Reached target Unmount All Filesystems.
[  OK  ] Stopped File System Check …dev/disk/by-label/ubuntu-seed.[   15.057412] 
systemd[1]: systemd-fsck@dev-disk-by\x2dlabel-ubuntu\x2dseed.service: Succeeded.
[  OK  ] Stopped File System Check on /dev/vda2.
[   15.059534] systemd[1]: Stopped File System Check on /dev/disk/by-label/ubuntu-seed.
[   15.061468] systemd[1]: systemd-fsck@dev-vda2.service: Succeeded.
[   15.061563] systemd[1]: Stopped File System Check on /dev/vda2.
[   15.061943] systemd[1]: Removed slice system-systemd\x2dfsck.slice.
[  OK  ] Removed slice system-systemd\x2dfsck.slice.
[  OK  [   15.065132] systemd[1]: Reached target Shutdown.
] Reached target Shutdown.
[[   15.066266] systemd[1]: Reached target Final Step.
  OK  ] Reached target Final Step.
[   15.067423] systemd[1]: systemd-reboot.service: Succeeded.
[   15.067525] systemd[1]: Finished Reboot.
[  OK  ] Finished Reboot.
[  OK  [   15.068878] systemd[1]: Reached target Reboot.
] Reached target Reboot.
[   15.069971] systemd[1]: Shutting down.
[   15.222012] systemd-shutdown[1]: Syncing filesystems and block devices.
[   15.224196] systemd-shutdown[1]: Sending SIGTERM to remaining processes...
[   15.228022] systemd-journald[181]: Received SIGTERM from PID 1 (systemd-shutdow).
[   15.236541] systemd-shutdown[1]: Sending SIGKILL to remaining processes...
[   15.240902] systemd-shutdown[1]: Unmounting file systems.
[   15.243301] [397]: Remounting '/' read-only in with options 'size=4053428k,nr_inodes=1013357'.
[   15.246232] systemd-shutdown[1]: All filesystems unmounted.
[   15.247573] systemd-shutdown[1]: Deactivating swaps.
[   15.248804] systemd-shutdown[1]: All swaps deactivated.
[   15.250084] systemd-shutdown[1]: Detaching loop devices.
[   15.252992] systemd-shutdown[1]: All loop devices detached.
[   15.255293] systemd-shutdown[1]: Detaching DM devices.
[   15.256663] systemd-shutdown[1]: All DM devices detached.
[   15.257959] systemd-shutdown[1]: All filesystems, swaps, loop devices and DM devices detached.
[   15.261069] systemd-shutdown[1]: Syncing filesystems and block devices.
[   15.262293] systemd-shutdown[1]: Rebooting.
[   15.263287] Unregister pv shared memory for cpu 0
[   15.263298] Unregister pv shared memory for cpu 1
[   15.263304] Unregister pv shared memory for cpu 2
[   15.263306] Unregister pv shared memory for cpu 3
[   15.263308] Unregister pv shared memory for cpu 7
[   15.263317] Unregister pv shared memory for cpu 4
[   15.263318] Unregister pv shared memory for cpu 5
[   15.274274] Unregister pv shared memory for cpu 6
[   15.277379] reboot: Restarting system
[   15.278511] reboot: machine restart
```

Given the above, I think we should either stick with `/sbin/reboot` or we should switch to using `systemctl reboot`, but it appears that `/sbin/reboot` will hook up to the same stuff that `systemctl reboot` does as you can see from the serial log output that with `/sbin/reboot` we still get to the Reboot target.

Reply:
that is interesting because it means "man reboot" is wrong or confused. Or maybe we get a slightly different variant inside initramfs though that doesn't seem the case ...

Reply:
I ended up doing a bare `/sbin/reboot` call, and wait for up to 10 minutes if reboot returns with 0 exit status, and then panic if we still haven't been rebooted here.